aif360/algorithms/preprocessing/lfr_helpers/helpers.py

```diff
@@ -1,109 +1,46 @@
+def LFR_optim_objective(parameters, x_unpriveleged, x_priveleged, y_unpriveleged,
+                        y_priveleged, k=10, A_x=0.01, A_y=0.1, A_z=0.5, print_inteval=250):
 
-    LFR_optim_obj.iters += 1
-    Ns, P = data_sensitive.shape
-    Nns, _ = data_nonsensitive.shape
+    num_unpriveleged, features_dim = x_unpriveleged.shape
+    num_priveleged, _ = x_priveleged.shape
 
-    alpha0 = params[:P]
-    alpha1 = params[P : 2 * P]
-    w = params[2 * P : (2 * P) + k]
-    v = np.matrix(params[(2 * P) + k:]).reshape((k, P))
+    w = parameters[:k]
+    prototypes = parameters[k:].reshape((k, features_dim))
 
-    dists_sensitive = distances(data_sensitive, v, alpha1, Ns, P, k)
-    dists_nonsensitive = distances(data_nonsensitive, v, alpha0, Nns, P, k)
+    M_unpriveleged, x_hat_unpriveleged, y_hat_unpriveleged = get_xhat_y_hat(prototypes, w, x_unpriveleged)
 
-    M_nk_sensitive = M_nk(dists_sensitive, Ns, k)
-    M_nk_nonsensitive = M_nk(dists_nonsensitive, Nns, k)
+    M_priveleged, x_hat_priveleged, y_hat_priveleged = get_xhat_y_hat(prototypes, w, x_priveleged)
 
-    M_k_sensitive = M_k(M_nk_sensitive, Ns, k)
-    M_k_nonsensitive = M_k(M_nk_nonsensitive, Nns, k)
+    y_hat = np.concatenate([y_hat_unpriveleged, y_hat_priveleged], axis=0)
+    y = np.concatenate([y_unpriveleged.reshape((-1, 1)), y_priveleged.reshape((-1, 1))], axis=0)
 
-    L_z = 0.0
-    for j in range(k):
-        L_z += abs(M_k_sensitive[j] - M_k_nonsensitive[j])
+    L_x = np.mean((x_hat_unpriveleged - x_unpriveleged) ** 2) + np.mean((x_hat_priveleged - x_priveleged) ** 2)
+    L_z = np.mean(abs(np.mean(M_unpriveleged, axis=0) - np.mean(M_priveleged, axis=0)))
+    L_y = - np.mean(y * np.log(y_hat) + (1. - y) * np.log(1. - y_hat))
 
-    x_n_hat_sensitive, L_x1 = x_n_hat(data_sensitive, M_nk_sensitive, v, Ns, P, k)
-    x_n_hat_nonsensitive, L_x2 = x_n_hat(data_nonsensitive, M_nk_nonsensitive, v, Nns, P, k)
-    L_x = L_x1 + L_x2
+    total_loss = A_x * L_x + A_y * L_y + A_z * L_z
 
-    yhat_sensitive, L_y1 = yhat(M_nk_sensitive, y_sensitive, w, Ns, k)
-    yhat_nonsensitive, L_y2 = yhat(M_nk_nonsensitive, y_nonsensitive, w, Nns, k)
-    L_y = L_y1 + L_y2
+    if LFR_optim_objective.steps % print_inteval == 0:
+        print("step: {}, loss: {}, L_x: {},  L_y: {},  L_z: {}".format(
+            LFR_optim_objective.steps, total_loss, L_x,  L_y,  L_z))
+    LFR_optim_objective.steps += 1
 
-    criterion = A_x * L_x + A_y * L_y + A_z * L_z
+    return total_loss
 
-    if LFR_optim_obj.iters % print_inteval == 0:
-        print(LFR_optim_obj.iters, criterion)
 
-    if results:
-        return yhat_sensitive, yhat_nonsensitive, M_nk_sensitive, M_nk_nonsensitive
-    else:
-        return criterion
-LFR_optim_obj.iters = 0
+def get_xhat_y_hat(prototypes, w, x):
+    M = softmax(cdist(x, prototypes), axis=1)
+    x_hat = np.matmul(M, prototypes)
+    y_hat = np.clip(
+        np.matmul(M, w.reshape((-1, 1))),
+        1e-6,
+        0.999
```

Comment:
these values seem sort of arbitrary -- was there a reason for choosing these as opposed to machine epsilon or something?

Reply:
fixed.